### PR TITLE
Add thread sort settings

### PR DESCRIPTION
**Pull request by pfrazee**
As requested!

CleanShot 2023-09-18 at 18 01 03@2x

### CODE REVIEW

**Comment by emilyliu7321** [comment]
`Sort Replies By ... ` -> `Sort Replies`
`Replies always show by ... ` -> `Sort replies to the same post by: `
`Followed Users First` -> `Prioritize Your Follows`
`Set this setting to...` -> `Show replies by people you follow before all other replies.` (and keep the `Yes` on the toggle itself)

**Comment by Sininini** [comment]
Since you're poking at reply sorting may I suggest always keeping, or at least having an option to keep your own replies at the very top. It's annoying to have to scroll through lots of replies to find your own.

File: src/state/models/content/post-thread.ts [hunk 1]
```diff
@@ -296,13 +301,31 @@ function sortThread(item: MaybeThreadItem) {
       if (modScore(a.moderation) !== modScore(b.moderation)) {
         return modScore(a.moderation) - modScore(b.moderation)
       }
-      if (a.post.likeCount === b.post.likeCount) {
-        return b.post.indexedAt.localeCompare(a.post.indexedAt) // newest
-      } else {
-        return (b.post.likeCount || 0) - (a.post.likeCount || 0) // most likes
+      if (opts.threadFollowedUsersFirst) {
+        const af = a.post.author.viewer?.following
+        const bf = b.post.author.viewer?.following
```
Comment (estrattonbailey) [inline src/state/models/content/post-thread.ts:306]: Sick

File: src/state/models/content/post-thread.ts [hunk 2]
```diff
@@ -296,13 +301,31 @@ function sortThread(item: MaybeThreadItem) {
...
+      } else if (opts.threadDefaultSort === 'most-likes') {
+        if (a.post.likeCount === b.post.likeCount) {
+          return b.post.indexedAt.localeCompare(a.post.indexedAt) // newest
+        } else {
+          return (b.post.likeCount || 0) - (a.post.likeCount || 0) // most likes
+        }
+      } else if (opts.threadDefaultSort === 'random') {
+        return 0.5 - Math.random() // this is vaguely criminal but we can get away with it
       }
+      return b.post.indexedAt.localeCompare(a.post.indexedAt)
```
Comment (estrattonbailey) [inline src/state/models/content/post-thread.ts:326]: This default is technically `newest` and not `oldest` like it is in preferences, right?

File: src/state/models/ui/preferences.ts [hunk 3]
```diff
@@ -189,6 +194,21 @@ export class PreferencesModel {
...
       }
+      // check if thread sort order is set in preferences, then hydrate
+      if (
+        hasProp(v, 'threadDefaultSort') &&
+        typeof v.threadDefaultSort === 'string' &&
+        THREAD_SORT_VALUES.includes(v.threadDefaultSort)
+      ) {
+        this.threadDefaultSort = v.threadDefaultSort
+      }
+      // check if tread followed-users-first is enabled in preferences, then hydrate
```
Comment (estrattonbailey) [inline src/state/models/ui/preferences.ts:205]: Nit, don't need it
```suggestion
      // check if thread followed-users-first is enabled in preferences, then hydrate
```

File: src/view/screens/PreferencesThreads.tsx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+              Sort replies to the same post by:
+            </Text>
+            <View style={[pal.view, {borderRadius: 8, paddingVertical: 6}]}>
+              <RadioGroup
+                type="default-light"
+                items={[
+                  {key: 'oldest', label: 'Oldest replies first'},
+                  {key: 'newest', label: 'Newest replies first'},
+                  {key: 'most-likes', label: 'Most-liked replies first'},
+                  {key: 'random', label: 'Random (aka "Poster\'s Roulette")'},
```
Comment (estrattonbailey) [inline src/view/screens/PreferencesThreads.tsx:59]: Would be ideal to generate this from the `THREAD_SORT_VALUES` array

Comment (estrattonbailey) [reply src/view/screens/PreferencesThreads.tsx:59]: Also lawl Poster's Roulette

**Review by estrattonbailey — APPROVED** [review]
Couple nits, v nice

File: src/state/models/content/post-thread.ts (hunk 2 above)
Comment (pfrazee) [reply src/state/models/content/post-thread.ts:326]: yes but that should never hit unless something goes real wrong

**Pull request merged**